client/src/main.rs

```diff
@@ -0,0 +1,239 @@
+use clap::Parser;
+use rustyline::Editor;
+use std::net::SocketAddr;
+use std::{io, net};
+use tokio::net::TcpListener;
+
+use fuel_vm::consts::{VM_MAX_RAM, WORD_SIZE};
+use fuel_vm::prelude::{Breakpoint, ContractId};
+
+use fuel_debugger::{Command, Response};
+
+use fuel_debugger_client::{names, Client, Listener};
+
+#[derive(Parser, Debug)]
+pub struct Opt {
+    #[clap(long = "ip", default_value = "127.0.0.1", parse(try_from_str))]
+    pub ip: net::IpAddr,
+
+    #[clap(long = "port", default_value = "4001")]
+    pub port: u16,
+}
+
+#[tokio::main]
+async fn main() -> io::Result<()> {
+    let config = Opt::parse();
+
+    let mut rl = Editor::<()>::new();
+
+    let listener = Listener::new(TcpListener::bind((config.ip, config.port)).await?);
+
+    println!(
+        "Listening for connections at {:?}",
+        SocketAddr::from((config.ip, config.port))
+    );
+
+    loop {
+        let (mut client, addr) = listener.accept().await?;
+        println!("Connected (remote {:?})", addr);
+
+        loop {
+            match rl.readline(">> ") {
+                Ok(line) => {
+                    rl.add_history_entry(line.as_str());
+                    if let Err(err) = parse_and_run_command(&mut client, line.as_str()).await {
+                        if err.kind() != io::ErrorKind::Other {
+                            println!("Error: {:?}", err);
+                        }
+                        break;
+                    }
+                }
+                Err(err) => {
+                    println!("{:?}", err);
+                    break;
+                }
+            };
+        }
+        println!("Disconnected");
+    }
+}
+
+macro_rules! ensure_no_more_args {
+    ($it:expr) => {
+        if $it.next().is_some() {
+            println!("Too many arguments for this command");
+            return Ok(());
+        }
+    };
+}
+
+async fn parse_and_run_command(client: &mut Client, text: &str) -> io::Result<()> {
```

Comment:
This is not a very maintainable / idiomatic cli pattern for Rust. Things like help text and command parsing should be automatically handled by clap (which you're already using in other places).